narwhal/worker/src/handlers.rs

```diff
@@ -340,39 +217,59 @@ impl PrimaryToWorker for PrimaryReceiverHandler {
         // If first request timed out or was missing batches, try broadcasting to some others.
         // TODO: refactor this to retry forever unless RPC is canceled. This will require more
         // invasive changes to primary code and cancellation propagation support in anemo.
-        let (tx_batch_response, rx_batch_response) = oneshot::channel();
         let message = WorkerBatchRequest {
             digests: missing.iter().cloned().collect(),
         };
-        self.tx_request_batches_rpc
-            .send((
-                None,
-                Some(self.request_batches_retry_nodes),
-                message,
-                tx_batch_response,
-            ))
-            .await
-            .map_err(|e| anemo::rpc::Status::internal(e.to_string()))?;
-        let response = rx_batch_response
-            .await
-            .map_err(|e| anemo::rpc::Status::internal(e.to_string()))?;
-        if let Ok(response) = response {
-            for batch in response.into_body().batches {
-                if missing.remove(&batch.digest()) {
-                    // Only send batch to processor if we haven't received it already
-                    // from another source.
-                    if self.tx_batch_processor.send(batch).await.is_err() {
-                        // Assume error sending to processor means we're shutting down.
-                        return Err(anemo::rpc::Status::internal("shutting down"));
+        let names: Vec<_> = self
+            .worker_cache
+            .load()
+            .others_workers(&self.name, &self.id)
+            .into_iter()
+            .map(|(_, info)| info.name)
+            .collect();
+        let mut clients: Vec<_> = names
+            .choose_multiple(&mut rand::thread_rng(), self.request_batches_retry_nodes)
+            .filter_map(|name| network.peer(anemo::PeerId(name.0.to_bytes())))
+            .map(WorkerToWorkerClient::new)
+            .collect();
+        let mut handles: FuturesUnordered<_> = clients
+            .iter_mut()
+            .map(|client| {
+                time::timeout(
+                    self.request_batches_timeout,
+                    client.request_batches(message.clone()),
+                )
+            })
+            .collect();
+        while let Some(result) = handles.next().await {
+            match result {
+                // TODO: is there a not-icky way to dedupe this code w/ above without async closures?
```

Comment:
We can potentially factor this code out into an async method that takes as input the batches

Reply:
I thought about this but decided moving stuff into a separate top level function somewhere else would be a net loss for readability. When I change it back to retrying indefinitely (until client cancels the RPC), then it will hopefully be easy to dedupe it as one loop

Reply:
Removed the TODO for now